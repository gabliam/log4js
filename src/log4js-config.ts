import {
  PluginConfig,
  Value,
  Bean,
  CORE_CONFIG,
  inject,
  interfaces
} from '@gabliam/core';
import { log4js } from './log4js';
import * as Joi from 'joi';
import * as path from 'path';
import * as fs from 'fs';

@PluginConfig()
export class LoggerConfig {
  @Value('application.loggerConfigPath', Joi.string())
  loggerConfigPath: string | null;

  private configPath: string;

  constructor(@inject(CORE_CONFIG) coreConfig: interfaces.GabliamConfig) {
    this.configPath = coreConfig.configPath;
  }

  @Bean('logger')
  createLogger() {
    if (this.loggerConfigPath && fs.existsSync(this.loggerConfigPath)) {
      log4js.configure(this.loggerConfigPath);
    } else {
      const defaultFilePath = path.resolve(this.configPath, 'log4js.json');
      if (fs.existsSync(defaultFilePath)) {
        log4js.configure(defaultFilePath);
      }
    }
    return log4js;
  }
}
