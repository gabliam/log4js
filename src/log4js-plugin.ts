import {
  interfaces as coreInterfaces,
  Scan,
  inversifyInterfaces,
  Registry
} from '@gabliam/core';
import { log4js } from './log4js';

@Scan(__dirname)
export class Log4jsPlugin implements coreInterfaces.GabliamPlugin {
  async destroy(container: inversifyInterfaces.Container, registry: Registry) {
    return new Promise<void>(resolve => {
      log4js.shutdown(() => resolve());
    });
  }
}
