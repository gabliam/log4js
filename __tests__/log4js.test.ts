import { Log4jsPluginTest } from './log4js-plugin-test';
import { Service } from '@gabliam/core';
import { log4js } from '../src/index';
import * as path from 'path';
import * as sinon from 'sinon';

let appTest: Log4jsPluginTest;
const p = sinon.spy(process.stdout, 'write');

describe('without config folder', () => {
  beforeEach(async () => {
    appTest = new Log4jsPluginTest();
  });

  afterEach(async () => {
    await appTest.destroy();
    p.reset();
  });

  test('without config', async () => {
    @Service()
    class TestLog {
      public logger = log4js.getLogger(TestLog.name);

      test() {
        this.logger.info('Info');
        this.logger.debug('debug');
        this.logger.error('error');
        this.logger.fatal('fatal');
        this.logger.warn('warn');
      }
    }

    appTest.addClass(TestLog);
    await appTest.build();
    const testLog = appTest.gab.container.get(TestLog);
    testLog.test();
    expect(p.callCount).toMatchSnapshot();
  });
});

describe('with config folder', () => {
  beforeEach(async () => {
    appTest = new Log4jsPluginTest(
      path.resolve(__dirname, './fixtures/defaultFilePathConfig')
    );
  });

  afterEach(async () => {
    await appTest.destroy();
    p.reset();
  });

  test('default config', async () => {
    @Service()
    class TestLog {
      public logger = log4js.getLogger(TestLog.name);

      test() {
        this.logger.info('Info');
        this.logger.debug('debug');
        this.logger.error('error');
        this.logger.fatal('fatal');
        this.logger.warn('warn');
      }
    }

    appTest.addClass(TestLog);
    await appTest.build();
    const testLog = appTest.gab.container.get(TestLog);
    testLog.test();
    expect(p.callCount).toMatchSnapshot();
  });

  test('custom config', async () => {
    @Service()
    class TestLog {
      public logger = log4js.getLogger(TestLog.name);

      test() {
        this.logger.info('Info');
        this.logger.debug('debug');
        this.logger.error('error');
        this.logger.fatal('fatal');
        this.logger.warn('warn');
      }
    }

    appTest.addConf(
      'application.loggerConfigPath',
      '../customConfig/log4js-custom.json'
    );
    appTest.addClass(TestLog);
    await appTest.build();
    const testLog = appTest.gab.container.get(TestLog);
    testLog.test();
    expect(p.callCount).toMatchSnapshot();
  });
});
