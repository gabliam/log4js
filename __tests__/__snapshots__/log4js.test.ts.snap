// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with config folder custom config 1`] = `3`;

exports[`with config folder default config 1`] = `4`;

exports[`without config folder without config 1`] = `0`;
